TypeCobol/Compiler/Diagnostics/CrossChecker.cs

```diff
@@ -208,103 +228,174 @@ public static void CheckPicture(Node node, CommonDataDescriptionAndDataRedefines
             }
         }
 
-        public static DataDefinition CheckVariable(Node node, StorageArea storageArea)
+        public static DataDefinition CheckVariable(Node node, StorageArea storageArea, bool isReadStorageArea)
         {
             if (storageArea == null || !storageArea.NeedDeclaration)
                 return null;
 
             var area = storageArea.GetStorageAreaThatNeedDeclaration;
-            IEnumerable<DataDefinition> found;
-            var foundQualified = new List<KeyValuePair<string, DataDefinition>>();
-
             if (area.SymbolReference == null) return null;
             //Do not handle TCFunctionName, it'll be done by TypeCobolChecker
             if (area.SymbolReference.IsOrCanBeOfType(SymbolType.TCFunctionName)) return null;
+            //need to initialize the dictionaries before the search
+            if (isReadStorageArea && node.StorageAreaReadsDataDefinition == null)
+            {
+                node.StorageAreaReadsDataDefinition = new Dictionary<StorageArea, Tuple<string, DataDefinition>>();
+            }
+            if (!isReadStorageArea && node.StorageAreaWritesDataDefinition == null)
+            {
+                node.StorageAreaWritesDataDefinition = new Dictionary<StorageArea, Tuple<string,DataDefinition>>();
+            }
 
-            var isPartOfTypeDef = (node as DataDefinition) != null && ((DataDefinition) node).IsPartOfATypeDef;
-            foundQualified =
+            //search for existing data definitinon before constructing one
```

Comment:
Do we need to search for existing data?
As CheckVariable is only called when the Visitor begin node, it shouldn't be.
I've created #996 to investigate this..